### PR TITLE
First pass of removing tx.origin usage.

### PR DESCRIPTION
**Pull request by paulgoleary**
- Changed many usages of `tx.origin` to `msg.sender`. Quite a few likely have no possible bad effects because if a method is designed to be called directly then `tx.origin` and `msg.sender` are the same.
- The most significant (only?) place this was not possible was related to factory object creation especially WRT factories. Because the factories actually create the objects if we used `msg.sender` to set ownership the the factory would be the owner. In theory, the space could just set the owner to `msg.sender` - because it's the space methods that are actually called by the end-user - but this would require us to change the space code, which we can't do easily.
- There are some related issues with setting rights after creation.
- However - as noted in the comments - we are likely fine in these cases because there doesn't seem to be a legit or dangerous 'spoofing' attack here because it's the `tx.origin` account which ends up as the owner a/o rights holder.
- Didn't touch `lv_recording.sol` because I'm not familiar with it - likely needs to be updated as well.

This branches from where `elv-master` `develop` is currently pointing in contracts. We need to catch up a bit on our branches and such here...

### CODE REVIEW

File: ownable.sol [hunk 1]
```diff
@@ -25,6 +25,12 @@ contract Ownable {
     address public owner;
     address public contentSpace;
 
+    // TODO: don't know if we can easily change this to msg.sender...
+    //  since object in the content hierarchy are created through factories, msg.sender would end up being the factory
+    //  instead of the end-user as expected. it theory we could change the space to set the owner as msg.sender but
```
Comment (elv-marc) [inline ownable.sol:30]: We can deal with that by adding in the factory a check that (msg.sender == tx.origin) that would avoid having the creation hijacked by another object above

**Review by elv-marc — COMMENTED** [review]
The one place I was worried about is the publish of content object, but it seems it will work,

**Comment by paulgoleary** [comment]
I believe this is more or less final now.

There's really just 3 explicit uses of `tx.origin` now and I think they are all safe / justifiable:
- Used to set `creator` in `Ownable`. This one is pretty trivial and also very easy to change because `creator` is not used for anything in the system right now AFAIK. But since we *do* expect `tx.origin` to be the final owner it seems like it might be useful as a check on how an object was actually created.
- Object factory methods. The general pattern here now is for the factory creation as before. But due to the changes in `Ownable` - for the object created through factories - the owner is initially the factory itself. But factories now switch ownership to `tx.origin` after they perform any necessary init on the objects. I believe that using `tx.origin` here is safe and justified because it's hard to imagine a legit or valuable 'phishing' attack where the 'phishee' ends up as the owner of the object.
- The more problematic use of `tx.origin` was to facilitate the rights 'proxy' pattern, where we have helper functions on objects that can be called to give permissions to other account / wallets. This is problematic because the call chain eventually reaches the `AccessIndexor` and those calls are expected to behave correctly whether they are called 'directly' or proxied. The solution here seems to be for the `AccessIndexor` to behave slightly different when it knows it's being called from a proxy - a situation that is trivial to detect. This behavior seems safe and justifiable since when the rights are being set on the object being 'proxied through' it would make no sense for that object to behave in a way that breaks security guarantees. In particular, our default implementation - `setRights` not validates explicitly with `msg.sender` so can't be called through a spoofed chain like before.
  - To be more explicit: the phishing 'attack' looks like this: [rights owner A] -> [bad content object] -> [bad custom contract] -> [set rights on owner A obj]
  - At the point `setRights` is called, `msg.sender` is the rogue contract address so the 'proxying' is not allowed.

`lv_recording.go` also has several uses of `tx.origin` but I'm leaving that alone for now because 1) I'm not 100% sure that the code is used as-is and 2) I'm unfamiliar with it.
@elv-marc should likely take a look a/o comment.

File: base_content_space.sol [hunk 2]
```diff
@@ -135,23 +135,23 @@ contract BaseContentSpace is MetaObject, Container, UserSpace, NodeSpace, IKmsSp
     }
 
     function engageAccountLibrary() public returns (address) {
```
Comment (elv-marc) [inline base_content_space.sol]: I am pretty sure we can get rid of  "engageAccountLibrary()"

File: base_content_space.sol [hunk 3]
```diff
@@ -294,9 +294,13 @@ contract BaseFactory is Ownable {
 
     bytes32 public version ="BaseFactory20200316120700ML"; //class name (max 16), date YYYYMMDD, time HHMMSS and Developer initials XX
 
+    // TODO: similar issue as tx.origin in Ownable - can't use msg.sender here because it's the space. But as with ownable,
+    //  don't think there's a legit spoofing attack here because the spoofee ends up with the rights.
     function createContentType() public returns (address) {
         address newType = (new BaseContentType(msg.sender));
```
Comment (elv-marc) [inline base_content_space.sol:304]: if we add on line 300:
require (msg.sender == contenSpace);

we ensure the call is not coming from alien object,

Moreover if we modify the BaseContentSpace method as follow:
function createContentType() public returns (address) {
        require(msg.sender == tx.origin);
        address contentTypeAddress = BaseFactory(factory).createContentType();
        emit CreateContentType(contentTypeAddress);
        return contentTypeAddress;
    }
Then we have ensured that the tx.origin in BaseContentType is the actual end user as intended.

We can use the same pattern for the other factories, except maybe the BaseContentObject as I think it has one more indirection.

File: base_content_space.sol [hunk 4]
```diff
@@ -320,9 +324,12 @@ contract BaseGroupFactory is Ownable {
 
     bytes32 public version ="BaseGroupFactory20200316120800ML"; //class name (max 16), date YYYYMMDD, time HHMMSS and Developer initials XX
 
+    // see note on BaseFactory
     function createGroup() public returns (address) {
         address newGroup = (new BaseAccessControlGroup(msg.sender));
```
Comment (elv-marc) [inline base_content_space.sol:338]: same comment as for Content Type above

File: base_library.sol [hunk 5]
```diff
@@ -75,21 +75,6 @@ contract BaseLibrary is MetaObject, Container {
         return canEdit();
     }
```
Comment (elv-marc) [inline base_library.sol:77]: We should modify the content creation as follow in base_library.sol:

function createContent(address content_type) public  returns (address) {
        require(msg.sender == tx.origin);
        address content = IFactorySpace(contentSpace).createContent(address(this), content_type);
        emit ContentObjectCreated(content, content_type, contentSpace);
        return content;
    }

File: base_content_space.sol [hunk 6]
```diff
@@ -360,10 +370,13 @@ contract BaseContentFactory is Ownable {
 
     bytes32 public version ="BaseCtFactory20200422180700ML"; //class name (max 16), date YYYYMMDD, time HHMMSS and Developer initials XX
 
+    // see note on BaseFactory re tx.origin
     function createContent(address lib, address content_type) public  returns (address) {
         Container libraryObj = Container(lib);
```
Comment (elv-marc) [inline base_content_space.sol:395]: insert before line 375:
require(msg.sender == contentSpace)

File: base_content_space.sol [hunk 7]
```diff
@@ -294,9 +294,13 @@ contract BaseFactory is Ownable {
...
+    // TODO: similar issue as tx.origin in Ownable - can't use msg.sender here because it's the space. But as with ownable,
+    //  don't think there's a legit spoofing attack here because the spoofee ends up with the rights.
     function createContentType() public returns (address) {
         address newType = (new BaseContentType(msg.sender));
-        BaseContentType(newType).setRights(tx.origin, 0 /*TYPE_SEE*/, 2 /*ACCESS_CONFIRMED*/); // register library in user wallet
+        BaseContentType theType = BaseContentType(newType);
+        theType.setRights(tx.origin, 0 /*TYPE_SEE*/, 2 /*ACCESS_CONFIRMED*/); // register library in user wallet
+        theType.transferOwnership(tx.origin);
         return newType;
     }
```
Comment (elv-marc) [inline base_content_space.sol:310]: In base_content_space.sol, we can modify the BaseContentSpace implementation of createContent:

 function createContent(address lib, address content_type) public returns (address) {
        require(msg.sender == lib);
        address contentAddress = BaseContentFactory(contentFactory).createContent(lib, content_type);
        emit CreateContent(contentAddress);
        return contentAddress;
    }


This way, we have ensured that the end user called the function on the lib, that the lib called the function on the content space and that the content space called the method on the factory, the the tx.origin is guaranteed to be the original end user.

File: base_content_space.sol (hunk 2 above)
Comment (paulgoleary) [reply base_content_space.sol]: Done

File: ownable.sol (hunk 1 above)
Comment (paulgoleary) [reply ownable.sol:30]: I think that check would need to be in the space to be effective.

Comment (paulgoleary) [reply ownable.sol:30]: Made changes in space.

File: base_content_space.sol (hunk 3 above)
Comment (paulgoleary) [reply base_content_space.sol:304]: Done

File: base_library.sol (hunk 5 above)
Comment (paulgoleary) [reply base_library.sol:77]: Done

File: base_content_space.sol (hunk 7 above)
Comment (paulgoleary) [reply base_content_space.sol:310]: Done

File: base_access_control_group.sol [hunk 8]
```diff
@@ -113,6 +113,14 @@ contract BaseAccessControlGroup is MetaObject, CounterObject, AccessIndexor, Edi
         tenant = _tenantAddr;
     }
 
+    // overload ...
```
Comment (elv-serban) [inline base_access_control_group.sol:116]: What does this comment mean here?

File: base_content_space.sol [hunk 9]
```diff
@@ -294,9 +294,17 @@ contract BaseFactory is Ownable {
 
     bytes32 public version ="BaseFactory20200316120700ML"; //class name (max 16), date YYYYMMDD, time HHMMSS and Developer initials XX
 
+    constructor(address _spaceAddr) public {
+        contentSpace = _spaceAddr;
+    }
+        // TODO: similar issue as tx.origin in Ownable - can't use msg.sender here because it's the space. But as with ownable,
+    //  don't think there's a legit spoofing attack here because the spoofee ends up with the rights.
```
Comment (elv-serban) [inline base_content_space.sol:301]: Is this comment still valid?

**Review by elv-serban — APPROVED** [review]
Looks good
I have some comments on the 'comments'  - there are a couple of TODO comments left (we can handle separately)

**Pull request merged**